state/allwatcher.go

```diff
@@ -347,27 +359,25 @@ func (u *backingUnit) updated(st *State, store *multiwatcherStore, id string) er
 	if u.CharmURL != nil {
 		info.CharmURL = u.CharmURL.String()
 	}
+	unit, err := st.Unit(u.Name)
+	if err != nil {
+		return errors.Annotatef(err, "cannot get unit %q", u.Name)
+	}
 	oldInfo := store.Get(info.EntityId())
 	if oldInfo == nil {
 		logger.Debugf("new unit %q added to backing state", u.Name)
 		// We're adding the entry for the first time,
 		// so fetch the associated unit status and opened ports.
-		unitStatus, agentStatus, err := unitAndAgentStatus(st, u.Name)
+
+		// process Unit and workload status.
+		err := u.processUnitAndAgentStatus(unit, info)
```

Comment:
We need to get the agent version info always since any update could result in a version change.